A46-xds-nack-semantics-improvement.md

```diff
@@ -0,0 +1,122 @@
+in the spec, this implies that the client must reject (i.e., not
+actually use) even the valid resources from that response, since that
+would allow the server to know the client's actual state (although see
+[Rationale](#rationale) below).  That is the behavior that gRPC currently
+implements.
+
+However, this approach causes problems in a case where a single resource
+is invalid and it causes the client to ignore *all* of the resources in
+the update, especially for LDS and CDS, where the server must send all
+resources in every response.  For example, if there is a single invalid
+Cluster resource in a CDS response, all of the Cluster resources will
+be rejected.  If this happens on the first CDS response after a client
+starts up (i.e., when the client has not already accepted previous
+versions of the CDS resources that it can continue to use), that will
+cause the client to have no valid CDS resources at all, which means
+that the problem will prevent *all* clusters from functioning instead of
+affecting only the invalid resource.  This is particularly problematic
+now that gRPC shares its XdsClient between channels, because a single
+invalid Cluster resource can basically cause all of the client's channels
+to stop working all at once.
+
+This behavior makes it challenging for xDS servers to safely deploy
+changes in environments in which the clients are not centrally controlled.
+For example, older gRPC clients support only the `ROUND_ROBIN`
+LB policy, as per [gRFC A27](A27-xds-global-load-balancing.md),
+but newer clients now support the `RING_HASH` policy, as per [gRFC
+A42](A42-xds-ring-hash-lb-policy.md).  If a Cluster resource specifies
+an unsupported LB policy, clients will consider the resource invalid,
+which will cause them to NACK the response.  So if an xDS server cannot
+be sure that all of its clients have been upgraded to a version that
+supports the `RING_HASH` policy, then it cannot safely send a Cluster
+resource configuring that policy, because that change would cause older
+clients to stop functioning.
+
+Note that this document is not attempting to solve the general problem
+that an individual resource that sets a supported field to an unsupported
+value will be considered invalid; that behavior is intentional, and there
+is nothing we can do to prevent the need for clients to be upgraded to
+use a configuration resource that requires features that they do not yet
+support.  However, this document *is* intending to solve the problem of
+that one invalid resource causing other *valid* resources to be ignored.
+
+This document proposes a behavior change to address this problem.
+
+### Related Proposals: 
+* [gRFC A27: xDS-Based Global Load Balancing](A27-xds-global-load-balancing.md)
+* [gRFC A42: xDS Ring Hash LB Policy](A42-xds-ring-hash-lb-policy.md)
+
+## Proposal
+
+We will change gRPC's behavior such that when a response is NACKed, gRPC
+will still use all valid resources from the response; it will ignore
```

Comment:
How do we update CSDS?

The invalid resources can be updated to the most recent version, and NACK

How the valid resources?
- Do we also update them to the most recent version, and NACK
- Or keep the previous version and ACK
- Or the most recent version and ACK???

Reply:
In CSDS, we report status on a per-resource basis, so it's fine for each resource to be in a different state.  The invalid resources will show status NACKED, and the valid resources can show ACKED with the new version.

Reply:
What if an invalid resource is so broken that is doesn't have a resource name to update CSDS, e.g. the ANY type message of that particular resource can not be parsed?

Reply:
> What if an invalid resource is so broken that it doesn't have a resource name to update CSDS

Oh, I think in that case we don't update CSDS for that resource even without the change.

Reply:
Yes, that's correct.  If we can't even find the resource's name, we have no way to report the problem via CSDS, but that's not affected by this proposal.